Order Creation: Show shipping amount field and currency symbol without fake text field

rachelmcr opened the pull request:
## Description

As discussed in https://github.com/woocommerce/woocommerce-ios/pull/6222#pullrequestreview-887222081, the fake `TextField` on the Shipping screen in Order Creation is confusing because it doesn't show a cursor when the field has focus.

This removes that fake field and uses the `BindableTextfield` as the main, visible field on screen with a currency symbol next to the field. (Internal discussion about this approach: p1645206501136169-slack-C02KUCFCSFP)

## Changes

* Replaces the hidden `BindableTextfield` and fake `TitleAndTextFieldRow` in the `ShippingLineDetails` view with a visible `Text` title and `BindableTextfield`, which directly formats the provided input.
* Adds a view modifier `addingCurrencySymbol(_:on:)` that positions the provided symbol to the left or right of the view it modifies. This is added to the `BindableTextfield` to show a currency symbol in the expected position according to the store currency settings.
* Updates the view model with a placeholder (that doesn't contain any currency symbol), currency symbol, and currency position. Removes the unused `formattedAmount` property.
* Updates unit tests to check the `amount` property itself now that the `formattedAmount` is not needed.

## Testing

1. Optional: Change your store's currency options in wp-admin, under WooCommerce > Settings.
2. Build and run the app.
3. Make sure Order Creation is enabled under Settings > Experimental Features.
4. Go to the Orders tab, tap the + button, and select "Create order."
5. Add a product to the order.
6. Select "Add Shipping."
7. Notice the Amount row shows your store's currency symbol, with the correct currency position and a `0` placeholder.
8. Enter a shipping amount and confirm it uses the correct decimal separator and number of decimals, according to your store's settings.
9. Save the shipping and confirm the correct amount is saved and reflected on the order.

## Screenshots

Placeholder|Amount with US currency settings|Amount with alternate currency settings
-|-|-
![Simulator Screen Shot - iPhone 13 Pro - 2022-02-22 at 00 12 57](https://user-images.githubusercontent.com/8658164/155043024-3507293e-1111-48b8-85b5-ae67189abb5b.png)|![Simulator Screen Shot - iPhone 13 Pro - 2022-02-22 at 00 12 51](https://user-images.githubusercontent.com/8658164/155043017-9a49071c-901e-4ad1-aa1a-36ed751fe5ea.png)|![Simulator Screen Shot - iPhone 13 Pro - 2022-02-22 at 00 18 40](https://user-images.githubusercontent.com/8658164/155043035-ba6eb926-752f-41f8-aa33-fb4597e7ab25.png)


https://user-images.githubusercontent.com/8658164/155042947-e6a4c86f-3eca-4e93-b81a-fe97e0a2769a.mp4





## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rachelmcr committed:
Show shipping amount field and currency symbol without fake text field

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/ShippingLineDetails.swift:
```diff
@@ -26,26 +26,20 @@ struct ShippingLineDetails: View {
+                                        focusAmountInput = true
+                                    }
                             }
+                            .padding()
```
To keep the same frame height as in `TitleAndTextFieldRow` we need different settings:
https://github.com/woocommerce/woocommerce-ios/blob/1cbf4c1bcd8fc2f811a42c61edfe4c03cc2c5bb9/WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI%20Components/TitleAndTextFieldRow.swift#L49-L50

rachelmcr replied:
Good catch! Updated that in [5a8051d](https://github.com/woocommerce/woocommerce-ios/pull/6268/commits/5a8051de8736bd8d2b098cbec1362144b0cb6998).

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/ShippingLineDetails.swift:
```diff
@@ -102,6 +96,38 @@ struct ShippingLineDetails: View {
 
+/// Adds a currency symbol to the left or right of the provided content
+///
+private struct CurrencySymbol: ViewModifier {
```
Do you think it deserves to be public? Not necessary in this PR, when adding to an additional screen.

rachelmcr replied:
Yep, that makes sense. I defaulted to private here (especially since it was a bit experimental at first) but for reuse it could be public and we could move this perhaps to `WooCommerce/Classes/View Modifiers`.

ealeksandrov commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/ShippingLineDetails.swift:
```diff
@@ -102,6 +96,38 @@ struct ShippingLineDetails: View {
+                content
+                Text(symbol)
+                    .bodyStyle()
+            }
```
Can we handle different spacing cases? With `spacing: .zero` in `HStack` and additional spacer views.
<img width=350 src=https://user-images.githubusercontent.com/3132438/155110294-49c69418-bc72-4cdd-aa32-897b4bb79971.png>

rachelmcr replied:
Good idea! I made some changes in [9ca7b63](https://github.com/woocommerce/woocommerce-ios/pull/6268/commits/9ca7b632e2cfca887bf125dacfa463fd538db05d) but ended up with the implementation in [9626247](https://github.com/woocommerce/woocommerce-ios/pull/6268/commits/9626247bb77af1e152f691c7df576409cc1703fb) to make it work with the way the text field is constrained (so the currency symbol stays closer to the text field).

ealeksandrov reviewed (approved):
Looks really nice, thank you for spending additional effort on improving UX here!
I found no issues in text editing flow 🚀 

There are minor UI considerations + the only important issue is a lack of width constraint:
<img width=350 src=https://user-images.githubusercontent.com/3132438/155112213-82d53b4a-dd81-410e-8cb1-d8787bca65b4.png>

BTW, maybe should we limit max input, not just frame? (but frame needs to be clipped separately anyway, for Dynamic Type support)

rachelmcr committed:
Set constant heigh and padding for amount row

rachelmcr committed:
Handle different spacing cases for currency position

rachelmcr committed:
Constrain width of BindableTextfield

rachelmcr commented:
Thanks for the review! I've addressed the issues and tested each currency position as well as the Simple Payment screen, and everything seems to be working as expected now. I've left this as a private modifier for the `ShippingLineDetails` view for now. I'll merge it as-is now, and we can make it public (and make any needed improvements) when we use it on another screen.

> the only important issue is a lack of width constraint

Thanks for noticing this! This turned out to be a little trickier than I expected, because `BindableTextfield` is a wrapped `UITextField` and the expected SwiftUI approaches for constraining its width didn't work for me. I found [this approach](https://stackoverflow.com/questions/59192194/swiftui-state-change-in-uitextfield-causes-the-framewidth-to-grow-uncontrolled/59193838#59193838) worked, along with some tweaks to the `ShippingLineDetails` layout:

/|Placeholder with Space|No space, large number|With space, large number
-|-|-|-
Left|![Shipping-LeftwithSpace-Placeholder](https://user-images.githubusercontent.com/8658164/155160389-eb6e8480-f36f-4b82-8ec3-b06b8095ddcd.png)|![Shipping-Left](https://user-images.githubusercontent.com/8658164/155159301-1ea405e3-64b6-4aec-86c6-b62ac8019396.png)|![Shipping-LeftwithSpace](https://user-images.githubusercontent.com/8658164/155159292-8ea9efc3-d0c4-4f38-b979-ba522d3b28e8.png)
Right|![Shipping-RightwithSpace-Placeholder](https://user-images.githubusercontent.com/8658164/155160459-793a2523-161f-40e3-bf98-eb33101c6fdb.png)|![Shipping-Right](https://user-images.githubusercontent.com/8658164/155159324-8a5cd5f5-89e4-442e-a163-03dcdb328a42.png)|![Shipping-RightwithSpace](https://user-images.githubusercontent.com/8658164/155159338-1bb872f7-ea1f-486d-b066-d5d86079a38c.png)

**Dynamic text:**

<img src="https://user-images.githubusercontent.com/8658164/155159353-1a108d55-6701-4912-a2dd-d20c9cda5219.png" width="300px">


> BTW, maybe should we limit max input, not just frame?

I'm not too worried about limiting the input separately from the frame. I can't think of a real-life scenario where that's likely to be an issue, so I'll leave it without any limit for now, but we could always add a limit if we decide there's a need for it.
